zypp/target/rpm/RpmDb.cc

```diff
@@ -1160,60 +1172,80 @@ namespace
     rpmts ts = ::rpmtsCreate();
     ::rpmtsSetRootDir( ts, root_r.c_str() );
     ::rpmtsSetVSFlags( ts, RPMVSF_DEFAULT );
-
-    rpmQVKArguments_s qva;
-    memset( &qva, 0, sizeof(rpmQVKArguments_s) );
-#ifndef HAVE_RPM_VERIFY_TRANSACTION_STEP
-    // Legacy: In rpm >= 4.15 qva_flags symbols don't exist
-    // and qva_flags is not used in signature checking at all.
-    qva.qva_flags = (VERIFY_DIGEST|VERIFY_SIGNATURE);
-#else
+#ifdef HAVE_RPM_VERIFY_TRANSACTION_STEP
     ::rpmtsSetVfyFlags( ts, RPMVSF_DEFAULT );
 #endif
+
     RpmlogCapture vresult;
     LocaleGuard guard( LC_ALL, "C" );	// bsc#1076415: rpm log output is localized, but we need to parse it :(
+    static rpmQVKArguments_s qva = ([](){ rpmQVKArguments_s qva; memset( &qva, 0, sizeof(rpmQVKArguments_s) ); return qva; })();
     int res = ::rpmVerifySignatures( &qva, ts, fd, path_r.basename().c_str() );
     guard.restore();
 
     ts = rpmtsFree(ts);
     ::Fclose( fd );
 
-    // results per line...
-    //     Header V3 RSA/SHA256 Signature, key ID 3dbdc284: OK
-    //     Header SHA1 digest: OK (a60386347863affefef484ff1f26c889373eb094)
-    //     V3 RSA/SHA256 Signature, key ID 3dbdc284: OK
-    //     MD5 digest: OK (fd5259fe677a406951dcb2e9d08c4dcc)
-    //
-    std::vector<std::string> lines;
-    str::split( vresult, std::back_inserter(lines), "\n" );
+    // Check the individual signature/disgest results:
+
+    // To.map back known result strings to enum, everything else is CHK_ERROR.
+    typedef std::map<std::string_view,RpmDb::CheckPackageResult> ResultMap;
+    static const ResultMap resultMap {
+      { "OK",         RpmDb::CHK_OK },
+      { "NOKEY",      RpmDb::CHK_NOKEY },
+      { "BAD",        RpmDb::CHK_FAIL },
+      { "UNKNOWN",    RpmDb::CHK_NOTFOUND },
+      { "NOTRUSTED",  RpmDb::CHK_NOTTRUSTED },
+      { "NOTFOUND",   RpmDb::CHK_NOTFOUND },
+    };
+    auto getresult = []( const ResultMap & resultMap, ResultMap::key_type key )->ResultMap::mapped_type {
+      auto it = resultMap.find( key );
+      return it != resultMap.end() ? it->second : RpmDb::CHK_ERROR;
+    };
+
+    // To track the signature states we saw.
     unsigned count[7] = { 0, 0, 0, 0, 0, 0, 0 };
 
-    for ( unsigned i = 1; i < lines.size(); ++i )
+    // To track the kind off sigs we saw.
+    enum Saw {
+      SawNone          = 0,
+      SawHeaderSig     = (1 << 0),  // Header V3 RSA/SHA256 Signature, key ID 3dbdc284: OK
+      SawHeaderDigest  = (1 << 1),  // Header SHA1 digest: OK (a60386347863affefef484ff1f26c889373eb094)
+      SawPayloadDigest = (1 << 2),  // Payload SHA256 digest: OK
+      SawSig           = (1 << 3),  // V3 RSA/SHA256 Signature, key ID 3dbdc284: OK
+      SawDigest        = (1 << 4),  // MD5 digest: OK (fd5259fe677a406951dcb2e9d08c4dcc)
+    };
+    unsigned saw = SawNone;
+
+    static const str::regex rx( "^ *(Header|Payload)? .*(Signature, key|digest).*: ([A-Z]+)" );
```

Comment:
Probably not really a problem, but this would also match non existant combinations of (Header|Payload) and (Signature, key|digest)